### PR TITLE
Add rating field to Book

### PR DESCRIPTION
**Pull request by 592363789**
the simple structure of rate

### CODE REVIEW

File: src/main/java/seedu/address/model/book/Rate.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
+package seedu.address.model.book;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Represents a book's rate.
+ * Guarantees: immutable.
+ */
+public class Rate {
```
Comment (takuyakanbr) [inline src/main/java/seedu/address/model/book/Rate.java]: Rename this class to `Rating`. (Shift+F6 or Refactor -> Rename)
Don't forget to change variables and descriptions to `rating`.

File: src/main/java/seedu/address/model/book/Rate.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+    public final String rate;
```
Comment (takuyakanbr) [inline src/main/java/seedu/address/model/book/Rate.java]: This should be an `int` between -1 to 5 (both inclusive).

File: src/main/java/seedu/address/model/book/Rate.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+    /**
+     * Constructs a {@code Description}.
+     *
+     * @param rate A book rate.
+     */
+    public Rate(String rate) {
```
Comment (takuyakanbr) [inline src/main/java/seedu/address/model/book/Rate.java]: This should take in an `int` `rating`.
Check the validity of the input using `AppUtil.checkArgument`, in a similar way to [this](https://github.com/nus-cs2103-AY1718S2/addressbook-level4/blob/master/src/main/java/seedu/address/model/person/Name.java).

File: src/main/java/seedu/address/network/api/google/BookDeserializer.java [hunk 4]
```diff
@@ -100,6 +101,7 @@ public void setVolumeInfo(JsonVolumeInfo volumeInfo) {
         private String publisher = "";
         private String publishedDate = "";
         private String description = "";
+        private String rate = "";
```
Comment (takuyakanbr) [inline src/main/java/seedu/address/network/api/google/BookDeserializer.java]: Remove this field.

File: src/main/java/seedu/address/network/api/google/BookDeserializer.java [hunk 5]
```diff
@@ -46,7 +46,8 @@ public Book deserialize(JsonParser jp, DeserializationContext ctxt) throws IOExc
         return new Book(new Gid(root.id), isbn,
                 BookDataUtil.getAuthorSet(volumeInfo.authors), new Title(volumeInfo.title),
                 getCategorySet(volumeInfo.categories), getDescription(volumeInfo.description),
-                new Publisher(volumeInfo.publisher), new PublicationDate(volumeInfo.publishedDate));
+                new Rate(volumeInfo.rate), new Publisher(volumeInfo.publisher),
```
Comment (takuyakanbr) [inline src/main/java/seedu/address/network/api/google/BookDeserializer.java]: Use the default rating instead.

File: src/main/java/seedu/address/network/api/google/BookShelfDeserializer.java [hunk 6]
```diff
@@ -117,6 +119,7 @@ public void setVolumeInfo(JsonVolumeInfo volumeInfo) {
         private String publisher = "";
         private String publishedDate = "";
         private String description = "";
+        private String rate = "";
```
Comment (takuyakanbr) [inline src/main/java/seedu/address/network/api/google/BookShelfDeserializer.java]: Remove this field.

File: src/main/java/seedu/address/network/api/google/BookShelfDeserializer.java [hunk 7]
```diff
@@ -59,7 +59,9 @@ private void convertAndAddBook(BookShelf bookShelf, JsonVolume volume, Gid gid)
         Book book = new Book(gid, isbn,
                 BookDataUtil.getAuthorSet(volumeInfo.authors), new Title(volumeInfo.title),
                 BookDataUtil.getCategorySet(volumeInfo.categories), new Description(volumeInfo.description),
+                new Rate(volumeInfo.rate),
```
Comment (takuyakanbr) [inline src/main/java/seedu/address/network/api/google/BookShelfDeserializer.java]: Use the default rating instead.

File: src/main/java/seedu/address/storage/XmlAdaptedBook.java [hunk 8]
```diff
@@ -53,16 +56,36 @@ public XmlAdaptedBook() {}
...
+        this.publicationDate = publicationDate;
+        this.publisher = publisher;
+    }
+
+    public XmlAdaptedBook(String gid, String isbn, String title, String description, String rate,
                           List<XmlAdaptedAuthor> authors, List<XmlAdaptedCategory> categories,
                           String publisher, String publicationDate) {
         this.title = title;
         this.description = description;
+        this.rate = "-1";
```
Comment (takuyakanbr) [inline src/main/java/seedu/address/storage/XmlAdaptedBook.java]: Remove this line.

File: src/test/data/BookDeserializerTest/ValidResponse.json [hunk 9]
```diff
@@ -11,6 +11,7 @@
     "publisher": "ABC Publishing House",
     "publishedDate": "2018-03-10",
     "description": "This is a valid description.",
+    "rate": "-1",
```
Comment (takuyakanbr) [inline src/test/data/BookDeserializerTest/ValidResponse.json]: Remove this line.

File: src/test/data/BookDeserializerTest/ValidResponseNoDesc.json [hunk 10]
```diff
@@ -10,6 +10,7 @@
     ],
     "publisher": "ABC Publishing House",
     "publishedDate": "2018-03-10",
+    "rate": "-1",
```
Comment (takuyakanbr) [inline src/test/data/BookDeserializerTest/ValidResponseNoDesc.json]: Remove this line.

File: src/test/data/BookShelfDeserializerTest/ValidResponse.json [hunk 11]
```diff
@@ -15,6 +15,7 @@
         "publisher": "ABC Publishing House",
         "publishedDate": "2018-03-10",
         "description": "This is a valid description.",
+        "rate": "-1",
```
Comment (takuyakanbr) [inline src/test/data/BookShelfDeserializerTest/ValidResponse.json]: Remove this line.

File: src/test/data/BookShelfDeserializerTest/ValidResponse.json [hunk 12]
```diff
@@ -56,6 +57,7 @@
         "publisher": "ABC Publishing House",
         "publishedDate": "2018-04-10",
         "description": "This is a valid description.",
+        "rate": "-1",
```
Comment (takuyakanbr) [inline src/test/data/BookShelfDeserializerTest/ValidResponse.json]: Remove this line.

File: src/test/data/BookShelfDeserializerTest/ValidResponse.json [hunk 13]
```diff
@@ -97,6 +99,7 @@
         "publisher": "ABC Publishing House",
         "publishedDate": "2018-05-10",
         "description": "This is a valid description.",
+        "rate": "-1",
```
Comment (takuyakanbr) [inline src/test/data/BookShelfDeserializerTest/ValidResponse.json]: Remove this line.

File: src/test/data/BookShelfDeserializerTest/ValidResponseDuplicateBooks.json [hunk 14]
```diff
@@ -15,6 +15,7 @@
         "publisher": "ABC Publishing House",
         "publishedDate": "2018-03-10",
         "description": "This is a valid description.",
+        "rate": "-1",
```
Comment (takuyakanbr) [inline src/test/data/BookShelfDeserializerTest/ValidResponseDuplicateBooks.json]: Remove this line.

File: src/test/data/BookShelfDeserializerTest/ValidResponseDuplicateBooks.json [hunk 15]
```diff
@@ -56,6 +57,7 @@
...
```
Comment (takuyakanbr) [inline src/test/data/BookShelfDeserializerTest/ValidResponseDuplicateBooks.json]: Remove this line.

File: src/test/data/BookShelfDeserializerTest/ValidResponseDuplicateBooks.json [hunk 16]
```diff
@@ -97,6 +99,7 @@
...
```
Comment (takuyakanbr) [inline src/test/data/BookShelfDeserializerTest/ValidResponseDuplicateBooks.json]: Remove this line.

File: src/test/data/BookShelfDeserializerTest/ValidResponseNoDesc.json [hunk 17]
```diff
@@ -14,6 +14,7 @@
         ],
         "publisher": "ABC Publishing House",
         "publishedDate": "2018-03-10",
+        "rate": "-1",
```
Comment (takuyakanbr) [inline src/test/data/BookShelfDeserializerTest/ValidResponseNoDesc.json]: Remove this line.

File: src/test/data/BookShelfDeserializerTest/ValidResponseNoDesc.json [hunk 18]
```diff
@@ -54,6 +55,7 @@
         ],
         "publisher": "ABC Publishing House",
         "publishedDate": "2018-04-10",
+        "rate": "-1",
```
Comment (takuyakanbr) [inline src/test/data/BookShelfDeserializerTest/ValidResponseNoDesc.json]: Remove this line.

File: src/test/data/XmlBookShelfStorageTest/invalidAndValidBookBookShelf.xml [hunk 19]
```diff
@@ -4,6 +4,7 @@
     <books>
         <title>Artemis</title>
         <description>This is Artemis.</description>
+        <rate>Rate</rate>>
```
Comment (takuyakanbr) [inline src/test/data/XmlBookShelfStorageTest/invalidAndValidBookBookShelf.xml]: Remove the extra `>` at the end of the line. Do the same for the other `.xml` files you modified.

File: docs/DeveloperGuide.adoc [hunk 20]
```diff
@@ -978,18 +978,18 @@ Use case ends.
 
 1. User obtains a list of books by listing (<<Use case: UC1 - View my reading list, *UC1*>>)
  or searching (<<Use case: UC2 - Search for books,*UC2*>>).
-2. User requests to give a rating to a specific book in the list.
+2. User requests to give a rate to a specific book in the list.
```
Comment (takuyakanbr) [inline docs/DeveloperGuide.adoc]: Revert this change.

File: docs/DeveloperGuide.adoc [hunk 21]
```diff
@@ -978,18 +978,18 @@ Use case ends.
...
 3. Bibliotek shows a success message.
 +
 Use case ends.
 
 *Extensions*
 
 [none]
-* 2a. The given rating is out of range.
+* 2a. The given rate is out of range.
```
Comment (takuyakanbr) [inline docs/DeveloperGuide.adoc]: Revert this change.

File: docs/DeveloperGuide.adoc [hunk 22]
```diff
@@ -978,18 +978,18 @@ Use case ends.
...
 +
 [none]
-** 2a1. Bibliotek says that the rating is invalid and asks user to try again.
+** 2a1. Bibliotek says that the rate is invalid and asks user to try again.
```
Comment (takuyakanbr) [inline docs/DeveloperGuide.adoc]: Revert this change.

File: src/main/java/seedu/address/logic/parser/ParserUtil.java [hunk 23]
```diff
@@ -125,4 +126,23 @@ public static Description parseDescription(String description) {
         return description.isPresent() ? Optional.of(parseDescription(description.get())) : Optional.empty();
     }
 
+    /**
+     * Parses a {@code String rate} into a {@code Rating}.
```
Comment (takuyakanbr) [inline src/main/java/seedu/address/logic/parser/ParserUtil.java]: `Parses a {@code String rating} into a {@code Rating}.`

File: src/main/java/seedu/address/logic/parser/ParserUtil.java [hunk 24]
```diff
@@ -125,4 +126,23 @@ public static Description parseDescription(String description) {
...
+     * Leading and trailing whitespaces will be trimmed.
+     */
+    public static Rating parseRating(int rating) {
+        requireNonNull(rating);
+        return new Rating(rating);
+    }
+
+    /**
+     * Parses a {@code Optional<String> rate} into an {@code Optional<Rating>}
```
Comment (takuyakanbr) [inline src/main/java/seedu/address/logic/parser/ParserUtil.java]: rate -> rating

File: src/main/java/seedu/address/model/book/Rating.java [hunk 25]
```diff
@@ -0,0 +1,41 @@
+package seedu.address.model.book;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Represents a book's rate.
```
Comment (takuyakanbr) [inline src/main/java/seedu/address/model/book/Rating.java]: rate -> rating

File: src/main/java/seedu/address/model/book/Rating.java [hunk 26]
```diff
@@ -0,0 +1,41 @@
...
+ * Guarantees: immutable.
+ */
+public class Rating {
+
+    public final Integer rating;
+
+    /**
+     * Constructs a {@code Description}.
```
Comment (takuyakanbr) [inline src/main/java/seedu/address/model/book/Rating.java]: `Description` -> `Rating`

File: src/test/java/seedu/address/logic/commands/CommandTestUtil.java [hunk 27]
```diff
@@ -34,9 +35,12 @@
     public static final String VALID_CATEGORY_BABYLON = "Fiction";
     public static final String VALID_DESCRIPTION_ARTEMIS = "This is Artemis.";
     public static final String VALID_DESCRIPTION_BABYLON = "This is Babylon's Ashes.";
+    public static final String VALID_RATING_ARTEMIS = "This is Artemis.";
+    public static final String VALID_RATING_BABYLON = "This is Babylon's Ashes.";
     public static final String VALID_ISBN_ARTEMIS = "234910247";
     public static final String VALID_ISBN_BABYLON = "907161513";
 
+
```
Comment (takuyakanbr) [inline src/test/java/seedu/address/logic/commands/CommandTestUtil.java]: Remove this extra newline.

File: src/test/java/seedu/address/logic/commands/CommandTestUtil.java [hunk 28]
```diff
@@ -45,9 +49,12 @@
     public static final String CATEGORY_DESC_BABYLON = " " + PREFIX_CATEGORY + VALID_CATEGORY_BABYLON;
     public static final String DESCRIPTION_DESC_ARTEMIS = " " + PREFIX_DESCRIPTION + VALID_DESCRIPTION_ARTEMIS;
     public static final String DESCRIPTION_DESC_BABYLON = " " + PREFIX_DESCRIPTION + VALID_DESCRIPTION_BABYLON;
+    public static final String RATING_DESC_ARTEMIS = " " + PREFIX_RATING + VALID_RATING_ARTEMIS;
+    public static final String RATING_DESC_BABYLON = " " + PREFIX_RATING + VALID_RATING_BABYLON;
     public static final String ISBN_DESC_ARTEMIS = " " + PREFIX_ISBN + VALID_ISBN_ARTEMIS;
     public static final String ISBN_DESC_BABYLON = " " + PREFIX_ISBN + VALID_ISBN_BABYLON;
 
+
```
Comment (takuyakanbr) [inline src/test/java/seedu/address/logic/commands/CommandTestUtil.java]: Remove this extra newline.

File: src/test/java/seedu/address/logic/commands/CommandTestUtil.java [hunk 29]
```diff
@@ -34,6 +35,8 @@
     public static final String VALID_CATEGORY_BABYLON = "Fiction";
     public static final String VALID_DESCRIPTION_ARTEMIS = "This is Artemis.";
     public static final String VALID_DESCRIPTION_BABYLON = "This is Babylon's Ashes.";
+    public static final String VALID_RATING_ARTEMIS = "This is Artemis.";
```
Comment (takuyakanbr) [inline src/test/java/seedu/address/logic/commands/CommandTestUtil.java]:
```
public static final int VALID_RATING_ARTEMIS = 5;
public static final int VALID_RATING_BABYLON = 4;
```

File: src/main/java/seedu/address/model/book/Rating.java [hunk 30]
```diff
@@ -0,0 +1,41 @@
...
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Represents a book's rating.
+ * Guarantees: immutable.
+ */
+public class Rating {
+
+    public final Integer rating;
```
Comment (takuyakanbr) [inline src/main/java/seedu/address/model/book/Rating.java]: Use an `int` instead?

File: src/main/java/seedu/address/model/book/Rating.java [hunk 31]
```diff
@@ -0,0 +1,41 @@
...
+
+    /**
+     * Constructs a {@code Rating}.
+     *
+     * @param rating A book rating.
+     */
+    public Rating(Integer rating) {
```
Comment (takuyakanbr) [inline src/main/java/seedu/address/model/book/Rating.java]: Use an `int` instead?

File: src/main/java/seedu/address/storage/XmlAdaptedBook.java [hunk 32]
```diff
@@ -34,6 +35,8 @@
     @XmlElement(required = true)
     private String description;
     @XmlElement(required = true)
+    private Integer rating;
```
Comment (takuyakanbr) [inline src/main/java/seedu/address/storage/XmlAdaptedBook.java]: Use an `int` instead, if possible.

File: src/main/java/seedu/address/model/book/Rating.java (hunk 31 above)
Comment (592363789) [reply src/main/java/seedu/address/model/book/Rating.java]: Using int will cause some error in the following override methods, so I changed to Integer
